pgaudit.c

```diff
@@ -265,24 +277,31 @@ static bool statementLogged = false;
  * Respond to callbacks registered with MemoryContextRegisterResetCallback().
  * Removes the event(s) off the stack that have become obsolete once the
  * MemoryContext has been freed.  The callback should always be freeing the top
- * of the stack, but the code is tolerant of out-of-order callbacks.
+ * of the stack except events for cursor such as DECLARE CURSOR FOR SELECT.
  */
 static void
 stack_free(void *stackFree)
 {
-    AuditEventStackItem *nextItem = auditEventStack;
+    AuditEventStackItem *freeItem = (AuditEventStackItem *) stackFree;
+    AuditEventStackItem **nextItem = &auditEventStack;
 
     /* Only process if the stack contains items */
-    while (nextItem != NULL)
+    while (*nextItem != NULL)
     {
         /* Check if this item matches the item to be freed */
-        if (nextItem == (AuditEventStackItem *) stackFree)
+        if (*nextItem == freeItem)
         {
-            /* Move top of stack to the item after the freed item */
-            auditEventStack = nextItem->next;
+            /* Delete the item to be freed */
+            *nextItem = (*nextItem)->next;
 
-            /* If the stack is not empty */
-            if (auditEventStack == NULL)
+            /*
+             * If the stack is not empty or an item for cursor audit events has 
+             * been freed.
+             */
+            if (*nextItem == NULL ||
+                freeItem->auditEvent.commandTag == T_DeclareCursorStmt ||
+                freeItem->auditEvent.commandTag == T_FetchStmt ||
+                freeItem->auditEvent.commandTag == T_ClosePortalStmt)
```

Comment:
It is not clear to me why this is needed. Is this really related to your patch or is it a general issue that needs to be addressed?

Reply:
> I made comment and code updates in b3a6a05. Mostly comment editing and formatting, but there was a big simplification in the logging.

Thank you so much for your comment and code updates.

> I made comment and code updates in b3a6a05. Mostly comment editing and formatting, but there was a big simplification in the logging.

I understood. I'm looking into this...

> It is not clear to me why this is needed. Is this really related to your patch or is it a general issue that needs to be addressed?

I'll add the reason for that. I'm working on it...

Reply:
> I made comment and code updates in b3a6a05. Mostly comment editing and formatting, but there was a big simplification in the logging.

Thanks for your comment and code updates.

> This patch does not work when 8831cef is merged in and it will definitely break when the PG14 changes are merged in (#146). We'll need to try and address those changes before this can be committed.

I understood, I'm looking into this...

> It is not clear to me why this is needed. Is this really related to your patch or is it a general issue that needs to be addressed?

I'll add comment for this.

Reply:
This is related to my patch to get the STATEMENT_ID of cursor audit events incremented correctly.

In my patch, the stack_free() function has changed to take one item per pop, Also the stack_free() should always be freeing the top of the stack except events for cursor such as `DECLARE CURSOR FOR SELECT`.

https://github.com/pgaudit/pgaudit/pull/133/files#diff-732f45c2566567903ce7d398b6a6a35131e40d9eb81dfa4f37a2201c630884e2R279

Please also refer to change in stack behavior:

https://github.com/pgaudit/pgaudit/pull/133#issuecomment-766658965

So the variable substatementTotal should be reset in stack_free() function not only when the search reached the end of stack(auditEventStack==NULL), but also when any of cursor events `DECLARE CURSOR`, `FETCH`, `CLOSE` was popped from stack.

https://github.com/pgaudit/pgaudit/pull/133/files#diff-732f45c2566567903ce7d398b6a6a35131e40d9eb81dfa4f37a2201c630884e2R297

I tried to compare the Audit events that are retrieved before and after the code change when a cursor is executed as below:
```
begin;
declare cur1 cursor for select /* cursor */ * from test where id <= 10;
fetch 3 from cur1;
close cur1;
```

The result before the code change is as follows(notice: `FETCH` and `CLOSE CURSOR` have the same STATEMENT_ID):
```
postgres=# begin;
2021-06-13 08:08:14.708 JST [5150] LOG:  AUDIT: SESSION,1,1,MISC,BEGIN,,,begin;,<not logged>
BEGIN
postgres=*# declare cur1 cursor for select /* cursor */ * from test where id <= 10;
2021-06-13 08:08:14.747 JST [5150] LOG:  AUDIT: SESSION,2,1,READ,SELECT,,,declare cur1 cursor for select /* cursor */ * from test where id <= 10;,<not logged>
2021-06-13 08:08:14.747 JST [5150] LOG:  AUDIT: SESSION,2,2,READ,DECLARE CURSOR,,,declare cur1 cursor for select /* cursor */ * from test where id <= 10;,<not logged>
DECLARE CURSOR
postgres=*# fetch 3 from cur1;
2021-06-13 08:08:14.748 JST [5150] LOG:  AUDIT: SESSION,3,1,MISC,FETCH,,,fetch 3 from cur1;,<not logged>
 id
----
  1
  2
  3
(3 rows)

postgres=*# close cur1;
2021-06-13 08:08:14.770 JST [5150] LOG:  AUDIT: SESSION,3,2,MISC,CLOSE CURSOR,,,close cur1;,<not logged>
CLOSE CURSOR
```

While, the result after the code change is as follows(notice: `FETCH` and `CLOSE CURSOR` have the different STATEMENT_ID):
```
postgres=# begin;
2021-06-13 07:03:18.297 JST [4816] LOG:  AUDIT: SESSION,7,1,MISC,BEGIN,,,begin;,<not logged>
BEGIN
postgres=*# declare cur1 cursor for select /* cursor */ * from test where id <= 10;
2021-06-13 07:03:18.320 JST [4816] LOG:  AUDIT: SESSION,8,1,READ,SELECT,,,declare cur1 cursor for select /* cursor */ * from test where id <= 10;,<not logged>
2021-06-13 07:03:18.320 JST [4816] LOG:  AUDIT: SESSION,8,2,READ,DECLARE CURSOR,,,declare cur1 cursor for select /* cursor */ * from test where id <= 10;,<not logged>
DECLARE CURSOR
postgres=*# fetch 3 from cur1;
2021-06-13 07:03:18.348 JST [4816] LOG:  AUDIT: SESSION,9,1,MISC,FETCH,,,fetch 3 from cur1;,<not logged>
 id
----
  1
  2
  3
(3 rows)

postgres=*# close cur1;
2021-06-13 07:03:18.349 JST [4816] LOG:  AUDIT: SESSION,10,1,MISC,CLOSE CURSOR,,,close cur1;,<not logged>
CLOSE CURSOR
```

Reply:
OK, I see what you are doing now. I'm not sure I like it, though. Tracing down issues in the stack has always been one of our biggest problems and this has made stack management quite a bit more complex.

I think it would be best to wait until after PG14 is branched before putting this in master. That means it would not be part of the official PG14 release, but since we don't backpatch that may not be very important to you.

Reply:
Thanks, I fully agree with you, I also don't like this part. I'll remove it from this enhancement.

Reply:
Hi @dwsteele,
I removed unnecessary stack processing you pointed out in commit [234892e](https://github.com/pgaudit/pgaudit/pull/133/commits/234892ed1793b2bd5032795cdf7315b3b22ec2c3)
, and confirmed all tests passed.

ok         5886 ms

=====================
 All 1 tests passed.
=====================

Reply:
> I think it would be best to wait until after PG14 is branched before putting this in master. That means it would not be part
> of the official PG14 release, but since we don't backpatch that may not be very important to you.

Hi @dwsteele , 

Could you please tell me if this patch for row_count will be added into the official PostgreSQL release in the future?

Currently, even in cloud environments such as Amazon RDS PostgreSQL, there are many Audit requirements that require row_count, so we strongly hope that it will be added in the official PostgreSQL release.

This patch does not affect the existing logic to get Audit Events, and the existing test case also shows OK.
In addition, As you pointed out, the following extra processing was unnecessary and has been deleted, Thank you very much.
```
            if (* nextItem == NULL ||
                freeItem-> auditEvent.commandTag == T_DeclareCursorStmt ||
                freeItem-> auditEvent.commandTag == T_FetchStmt ||
                freeItem-> auditEvent.commandTag == T_ClosePortalStmt)
```

If it's difficult to add this patch into the officer PostgreSQL, let me know please, I will try to find another way to get row_count.

Best Regards,
Zhao

Reply:
With your change in 234892ed17 the tests no longer pass, see https://github.com/pgaudit/pgaudit/runs/3049812146?check_suite_focus=true. If I revert 234892ed17 then they pass. Reverting all changes to `stack_free()` leads to a different error.

I'm just still not sure why these changes to `stack_free()` are required at all. I feel they are compensating for an issue elsewhere, or at least something that could be coded better.

We're out of time to commit this for the PG14 release. We can continue to consider it for the PG15 release.

Reply:
Hi @dwsteele,

Could you please confirm my change in [eaf2b74](https://github.com/pgaudit/pgaudit/commit/eaf2b749cbf6bc567678b5a71818047cd9547d42).
I've fixed to get row_count without modifying the original stack functions.

Regards,
Zhao

Reply:
Yes, this looks much better.